### ISSUE TITLE
Can't login

### ISSUE THREAD

**Issue by Prog72** [hepcat72/node-red-contrib-life360#20]
I've just restarted Node-RED and am getting the following error repeatedly.

`Life360 server error logging in: 403 Error: Request failed with status code 403 undefined`

Is anyone else experiencing this?

**Comment by hepcat72** [comment]
Well, I'm traveling and my home VPN doesn't seem to be working correctly (I can "connect" to my VPN, but cannot ssh or VNC in), so I can't get into my pi to look at my logs. I can't even load node red's web interface...  I can say that as of about noon yesterday, it was still working, because my geofence automations were triggered, though I haven't disconnected from the Life360 server to try reconnecting... and I wouldn't want to try that when I'm not home.

I'll check after the holiday.

**Comment by Prog72** [comment]
Thanks for your reply.

Mine was working fine until I restarted Node-RED earlier today and that's when the problem started. Hopefully it's something I've done rather than a global issue but I would definintely not disconnect for now!

Happy travels.

**Comment by hepcat72** [comment]
I would google the error. If there's a global issue, the home assistant folks probably would have reported and solved it by now.

**Comment by aots74** [comment]
Having the same problem. Figured it might just be a url change like last time. Anyone figured it out yet?

**Comment by ljgonzalez1** [comment]
+1

**Comment by hepcat72** [comment]
If anyone wants to fix it, go ahead and submit a PR. I won't be able to get to it until next weekend.

**Comment by redpages** [comment]
Same here

**Comment by Prog72** [comment]
I did some digging and it appears the good people at Life360 are doing all they can to stop their service working with home automation. They've never officially supported it so I guess that's fair enough.

I now have a new geofencing solution using PhoneTrack for Android, Nextcloud running on a Raspberry Pi, and HTTP requests in Node-RED. It's not as elegant but it doesn't rely on a third party, so that's a positive.

@hepcat72 I'd like to take this opportunity to thank for your work on the Life360 node over the past few years. It's been very much appreciated.

**Comment by Prog72** [comment]
I did some digging and it appears the good people at Life360 are doing all they can to stop their service working with home automation. They've never officially supported it so I guess that's fair enough.

I now have a new geofencing solution using PhoneTrack for Android, Nextcloud running on a Raspberry Pi, and HTTP requests in Node-RED. It's not as elegant but it doesn't rely on a third party, so that's a positive.

@hepcat72 I'd like to take this opportunity to thank for your work on the Life360 node over the past few years. It's been very much appreciated.

**Comment by hepcat72** [comment]
Thanks. Looks like the Life360 folks may have implemented a major refactor. I don't necessarily think it was necessarily targeted the home automation community, but I only marched through the [HA issues page on it](https://github.com/home-assistant/core/issues/105816). They rewrote their interface.

Life360 sounds like they may just be upping their auth game?

I'll get home mid-week this week and look into it in more detail.

I myself have a separate geofence solution I use (the old Geofency app for iOS) that uses webhooks. It's not as reliable as Life360. Sometimes the webhooks fail. Life360 is more reliable than Geofency because it's not a single webhook call upon crossing a geofence boundary.

**Comment by nlanfran** [comment]
Same problem here since a couple of days, see below:

![Screenshot from 2023-12-24 14-09-43](https://github.com/hepcat72/node-red-contrib-life360/assets/26627254/b4db2f6b-7b2c-4d23-85dc-fd7fa566ab94)

**Comment by ljgonzalez1** [comment]
From Life360 support
> Thank you for reaching out to Life360 about using the cURL API. I'll be providing you an information to this.
>
> We understand the importance of using the Home Assistant program with Life360. However, we want to inform you that we no longer support home automation programs such as Home Assistant, Google Home, Alexa, IFTTT, and others, and these programs are no longer compatible with Life360.
>
> Currently, we have no plans to reintroduce compatibility with these programs. Nonetheless, we appreciate your feedback on this matter, and I will share it with the appropriate team.
>
> Have a safe day,
>
> Sarah
> Life360 Customer Care

I'm cancelling my subscription and changing to [owntracks](https://owntracks.org/)

**Comment by hepcat72** [comment]
So I spent a good part of today (and a bit yesterday) futzing with the code, and the conclusion I'm coming to (that you guys may have already arrived at) is that Life360 is using cloudflare to ostensibly prevent "attacks/hackers".  Whether they think the home automation community falls into this category or not, we're running afoul of their new security measures. In fact, it seems they've even buried the fact that there exists a free version/tier of their service.

I discovered a few different layers of what is happening.

First, they have tweaked their authentication interface.  There is now a mix of "v3" and "v4" URLs in their API, and even if you fix that, the fact that your installs have been querying the API with the old (invalid) URLs has gotten your accounts blocked by cloudflare (more on them below).  Changing your account password may or may not work around the block, that is, assuming the API has been updated - and I implemented (though I have not yet pushed) the same changes that the HA community claim fixes the issue).

Second, anyone who has enabled their 2FA feature appear to be more likely to not be able to work out the block.  There appears to be no way to disable 2FA once enabled either.  Some users have resorted to creating new accounts.

Third (and I haven't seen this mentioned anywhere else - I just encountered it on my own), their password requirements are now more stringent (and my old password did not meet the new requirements).

Fourth, there may (or may not) be a new limitation on number of devices logged into the same account.  Reports on this are mixed.  Some users have gotten around this by creating a new account using a google voice phone number and adding that duplicate account to circles.

This is not the first code issue I have run into with cloudflare.  All of the cell phone providers started using cloudflare (or cloud-whatever - I discovered there are multiple names referring to the same company) to start "spam filtering email-to-text features about a year or so ago.  I had been reliant on that feature to have my home automations message me, and clouflare, once I'd worked out they had blocked my email, claimed that a "another user" had reported my single-purpose-custom-email (that only ever emails my own (other) account) as "spam", which of course was bullshit.  They unblocked me, but it happened again - and the process for getting unblocked is time consuming and annoying.

I will revisit this node's code again to see what more I can do over the next week or 2, but I just started looking into [owntracks](https://owntracks.org/) (thanks @ljgonzalez1).  It looks quite interesting, particularly WRT privacy and control, (though not as easy to add friends to).  I was particularly impressed that they support ibeacons, which I also use.  I started reading the "booklet" to see about setting up a server, though it might take some time to get going.

**Comment by StoricU** [comment]
Have used Life360 for many years and the regularly change something so that it don't work..
Have also been looking for an alternative to Life360, but I just have not found any solution that has been as reliable (ironically enough) as Life360 when it comes to reporting correct.

**Comment by hepcat72** [comment]
> Have used Life360 for many years and the regularly change something so that it don't work..
> Have also been looking for an alternative to Life360, but I just have not found any solution that has been as reliable (ironically enough) as Life360 when it comes to reporting correct.

I agree. I've had a similar experience. But I've also dealt with cloud flare blocking before and it's rather difficult to deal with. I'm just not sure there is going to be a reliable solution moving forward. I will continue to try and figure this out, but I think I have the correct code right now in a branch and it's not working and I think it's because cloudflare has blocked me.

**Comment by hepcat72** [comment]
Perhaps later today, I'll push my branch and a volunteer can try installing it to see if they can get their install to work. I tried for most of the day yesterday and I cannot get around the 403 (with the message saying I'm blocked). I tried changing my password, which has worked for some others (using the HA L360 integration) though there are still others (like me) that can't seem to get around it. I think unless I reach out to Life360 and have them create a ticket with cloud flare to unblock me (like I did a couple times with T-Mobile regarding email-to-text about a year ago), I'm not going to be able to get past it.

**Comment by hepcat72** [comment]
I still have some reading to do on `owntracks`, but my first impression is that it has the potential to **not** be as reliable based on the following detail from their docs: They only ever communicate location changes to the server when the phone's OS tells `owntracks` that the phone has moved significantly.  Dependability will thus hinge on how they handle faults (i.e. what happens if there's no signal when it tries to communicate with the server).  If it keeps trying (say, once a minute) until it succeeds, then I would say that that would be better than `Life360` which uses polling.  On the server side, it supports MQTT, which is reliable in that fashion, but I'll have to keep reading to see if it's well implemented for robustness.

I just pushed a branch with changes that may or may not work for people.  The branch is [`auth2024`](https://github.com/hepcat72/node-red-contrib-life360/tree/auth2024).  Would anyone be willing to try it out?  It hasn't worked for me, but some users have had success (and some haven't) with these same changes that were implemented in `homeassistant`.  You may need to change your Life360 password to get it to work.  And as I understand it, if you have enabled `Life360`'s 2FA, you may run into a limitation where `Life360` only allows 1 device to query their server, so it may work briefly (until you open the Life360 app on your phone).  Users who have not enabled 2FA shouldn't encounter that issue, though in either case, YMMV.  Please report back what happens when you try out that branch.  Follow this general process:

1. Disable the Life360 server node
2. Stop node-red
3. `git clone git@github.com:hepcat72/node-red-contrib-life360.git`
4. `git fetch`
5. `git checkout auth2024`
6. Replace `node-red-contrib-life360` somewhere under `node_modules` with the `node-red-contrib-life360` directory you cloned/checked-out
7. Change your Life360 password
8. Start node-red back up
9. Update the password in the Life360 server node and re-enable
10. Deploy and see what happens
11. If you still see a 403, take a look in the node-red error log and paste the error you get here in this thread.

The branch only changes the `index.js` file.  You could take a look and just copy the changes to the corresponding file under `node_modules`.  Let me know if you want to try it, but need further instructions.

Rob

**Comment by StoricU** [comment]
Have briefly testet owntracks today, but it did not report me entering my home zone before I had been home for a while. This was with using the significant changes monitoring mode. I'm going to try the "Move" mode later.

However, I tested the auth2024 branch.
I'm getting this error in the debug list: "Life360 server error logging in: 500 Error: Request failed with status code 500 undefined"
Is there any other log with more information?

I have not enabled 2FA and I changed the password before testing.

**Comment by StoricU** [comment]
Don't know if this has anything to do with the error:
![image](https://github.com/hepcat72/node-red-contrib-life360/assets/50556267/360d7797-0a0f-4b79-8860-6484fe8eb700)

FYI, I had this error before doing the branch update.

**Comment by hepcat72** [comment]
Huh. I ran into the 500 error, but I thought I'd fixed that... Let me see if I can fix it.

**Comment by hepcat72** [comment]
OK.  Reproduced and addressed the 500 error.  Try pulling and try again.

The 500 error doesn't show up in the log, but my error (403) after this change does come out in the console's error log.  It prints the entire error object, including the server response.  I wouldn't leave it running that way forever.  It's a lot of text (if you get the error).

**Comment by StoricU** [comment]
Followed the same procedure  (https://github.com/hepcat72/node-red-contrib-life360/issues/20#issuecomment-1874346985) to update, and now I get a 403 error.

**Comment by hepcat72** [comment]
OK.  Thanks for testing.  I've been corresponding more with the guy who helps maintain the Life360 integration in HA.  I had pointed out that I had never enabled 2FA (and assumed my app was older since I don't have the option to enable it) and he said that the enabling of 2FA is automatic if you have a phone number on your account (which I do).  He said that there's a separate branch for Life360 in HA where someone implemented a way to add your phone number to the initial authentication and added a way to include a 2FA code in a subsequent auth call.  ATM, that's my best bet as to what's happening - why we're still getting a 403.

Can you confirm @StoricU, if you have a phone number in your Life360 account?  That would corroborate this theory.  If so, I will see about adding some code for this work-around, as unsustainable as that may be...

**Comment by StoricU** [comment]
I see, I can confirm that I have registered my phone number.

I have just configured owntracks with Tasker to switch between different update principles based on if the phone is connected to any WiFi, but I fear that this will drain the battery pretty quickly when out and about.

**Comment by hepcat72** [comment]
That's disappointing to hear.  Based on my initial reading of owntrack's booklet, I was optimistic that it would not be a battery drain, since it only communicates location when the OS tells owntracks that the phone has moved.  I.e., it's not a polling strategy.  Did you set up your own polling?  I haven't gotten very far yet with reading about owntracks, so I'll be looking forward to your assessment of how well it works.

**Comment by StoricU** [comment]
Of course it depends on what you need.
I primarily need it to detect when I'm arriving home to open the garage door.
Using the "Significant Changes" strategy it does not report entering the Home zone until I have parked. So it is fairly quick.. Just not quick enough.
That's why I have now set up the tasker to change the strategy to "Move" when not connected to wifi. So I will check tomorrow if that strategy is quick enough. And see how the battery holds up..

**Comment by hepcat72** [comment]
Can you increase the size of your geofence around your house?  That's what I did with Life360 (so that the AC would run for a bit before I arrive in the summer).

**Comment by StoricU** [comment]
The zone can't be to big, then it will interfere with my traveling routes during the day 🙈
But that is something I also did with Life360. The zones on HA and owntracks is the same size as in Life360 now.
Have to do some field studies =)

**Comment by hepcat72** [comment]
OK.  I was studying the code changes for the OTP branch in the HA repo and I got a basic bash function to work to determine the login method for your account.  Mine is "phone" (i.e. I need to use the OTP method).  Here's the (ugly) code if you want to try it yourself to confirm your account requires the OTP method:

```
#!/bin/bash

#Set variables
username360="your@email.here"
password360="yourlife360password"
phone360="yourphonenumber"
phonecountrycode360="1"  # USA
secret360="Y2F0aGFwYWNyQVBoZUtVc3RlOGV2ZXZldnVjSGFmZVRydVl1ZnJhYzpkOEM5ZVlVdkE2dUZ1YnJ1SmVnZXRyZVZ1dFJlQ1JVWQ=="
agent360="com.life360.android.safetymapd/KOKO/23.50.0 android/13"  # necessary
apiurl3="https://api-cloudfront.life360.com/v3"
apiurl4="https://api-cloudfront.life360.com/v4"
timeout=300


function loginMethod {
  echo "$(date +%s) INFO: requesting user login method"
  isemail=$(curl -s -X GET -H "Accept: application/json" -H "cache-control: no-cache" -H "user-agent: $agent360" -H "Authorization: Basic $secret360" "$apiurl3/users/lookup?email=$username360" | grep -c loginMethod)
  if [ $isemail -eq 0 ]; then
    isphone=$(curl -s -X GET -H "Accept: application/json" -H "cache-control: no-cache" -H "user-agent: $agent360" -H "Authorization: Basic $secret360" "$apiurl3/users/lookup?countryCode=$phonecountrycode360&phone=$phone360" | grep -c loginMethod)

    if [ $isphone -eq 0 ]; then
      echo "ERROR: No login method found"
      loginMethod=error
    else
      loginMethod=phone
    fi
  else
    loginMethod=email
  fi
  echo $loginMethod
}

loginMethod
```

The above is a bit verbose.  If your login method is phone(/OTP), then if you issue this `curl` command (after setting the variables):
```
curl -s -X GET -H "Accept: application/json" -H "cache-control: no-cache" -H "user-agent: $agent360" -H "Authorization: Basic $secret360" "$apiurl3/users/lookup?countryCode=$phonecountrycode360&phone=$phone360"
```
You will see something like this (note how devoid of info mine is, despite having an email and last name in my account):
```
{"user":{"firstName":"Rob","avatar":null,"email":null,"loginEmail":null,"id":null,"lastName":null},"loginMethod":"phone"}
```
If you use the email command (and your method is phone/OTP), you'll see something like this:
```
{"user":{"firstName":""}}
```
If you have something wrong in your get request, you'll get something like this:
```
{"errorMessage":"This login was not found","url":"\/v3\/users\/lookup.json?username=hepcat72@gmail.com","status":404}
```
It took me awhile to get this to work, so that's enough for today.  Next time I sit down with this, I'll see if I can send the request to get an OTP.

**Comment by StoricU** [comment]
Just ran your code and can confirm I require OTP.. 
Got the same result as you.

**Comment by hepcat72** [comment]
I was thinking about the 1-device limitation thing once you've used the OTP login method... A user in the HA thread said that they don't run into a limitation when they access the Life360 app on their iPad. There must be another way to access the location data: one way for the device being tracked and one way for devices on the same account that are not being tracked. Maybe it's just another user agent. Maybe it's another url. Don't know. But I had ready that the way the OTP interface was discerned was using some sort of software to watch the traffic while an android phone logged into Life360. Probably the same could be done for a device that's not one being tracked.

Just wanted to document that thought. I still plan to flesh out the process for the OTP login from HA.

**Comment by pantherale0** [comment]
@hepcat72 hello, the multiple device limitation is purely a software limitation. When you click the button to sign out it sends a delete request to a certain endpoint to end all existing user sessions. The access token itself used doesn't change between that and future requests and the API can support multiple devices.

**Comment by hepcat72** [comment]
Oh I know that it's a software limitation and that there must exist a way to interact with the API to do it. The question is, how to do it. Do you know?

I have a separate convo going with one of the HA devs who has worked on this. They developed an OTA auth interface to get logged in, but other users say that as soon as they open the app on their phone, they start getting 403s in the Life360 HA integration. What I'm thinking is, there must be a different query, different parameter, or different auth for connecting to only receive location data (not connect to send (and receive), which is what I infer the OTA login is doing (even if we're not sending location data).

**Comment by pantherale0** [comment]
See here: https://github.com/home-assistant/core/issues/106967#issuecomment-1876123190

I uploaded a text file with a list of all the API endpoints defined in the mobile app. There's no refresh endpoint that I can see.

Ideally you'll need to figure out where the 403 actually comes from by checking the content of the response, both the API itself and cloudflare use the 403 status code, but they can both mean entirely different things.

If it's from Life360 you'll generally have a JSON response that contains an error stating the Auth token is invalid.

If it's cloudflare you'll get the "this page is blocked", which is entirely different and means the request never reached life360s servers in the first place (could be page rules, WAF, custom rules or who knows what else life360 have configured in cloudflare).

**Comment by pantherale0** [comment]
There's also this spreadsheet doc that contains a whole bunch of random API endpoints from traffic sniffing:

https://cloud.hrvy.uk/s/McSWExgiJ9YxGMT

This is what you are looking for to end other sessions (DELETE request, no body):

![image](https://github.com/hepcat72/node-red-contrib-life360/assets/7346615/e030358a-7782-4990-9f51-d7c4da73ffa3)

**Comment by hepcat72** [comment]
I had read your efforts in the HA issue thread and I was thinking about the traffic sniffing. This is all kind of over my head, so maybe this is a pointless question, but when the traffic is being sniffed, would it make sense to use a device that is not one that is being tracked, e.g. an iPad? Maybe there would be some alternate endpoints discovered?

**Comment by pantherale0** [comment]
> I had read your efforts in the HA issue thread and I was thinking about the traffic sniffing. This is all kind of over my head, so maybe this is a pointless question, but when the traffic is being sniffed, would it make sense to use a device that is not one that is being tracked, e.g. an iPad? Maybe there would be some alternate endpoints discovered?

You could try, although I'm an android only household and with android you need a rooted device to run frida-server to disable SSL pinning otherwise you won't be able to see any of the traffic. I've reset the device multiple times and there's no difference in any of the web requests sent.

**Comment by StoricU** [comment]
Have now been testing out OwnTracks for a while. Sadly it's a battery killer.
I'm using Android, and for android the "Significant location change mode" the documentation states: " It uses a balanced power location request that gathers a new location fix every 15 minutes"
So it's no good for automation using zone entering/leaving.
Using "Move" is a serious battery killer, since it requests a location fix every 10s.
Really wonder how Life360 works, since it has been flawless for so long a time without draining the battery

**Comment by ctreischman** [comment]
I believe Significant location change is every 15 minutes or if you stop traveling.  One nice thing with OwnTracks is that a region radius seems to be unlimited.  So far with my testing I made 3 regions at my house (Home=200m, Home Near=1000m, and Home Far=40,000m).  When i enter Home Far after a longer trip it sets my AC back earlier.  When i enter Home Near it sets OwnTracks to Move mode until i finally get to Home.  Then Home adjusts garage door, security, and some lights.

Life360 radii seemed to be limited and I couldn't create the largest circle.  So I created separate circles N, S, E, and W of my house and had to get creative with some logic on entering/exiting those circles to control the AC.

**Comment by StoricU** [comment]
Interesting @ctreischman !
And that works? 
How do you set the updating rate based on regions?
Is that a OwnTracks feature?
Up to now I have been using Tasker to control location updates, based on wifi connected and wifi near, but it has not been as reliable as hoped. Some times tasker don't turn it back to significant location when connecting to a wifi, so then the battery drains.
And I live so close to the city and my "during the day" travelroute, that the home zone can be max 300. And a Home Near would not work, since it would trigger almost all the day.

**Comment by ctreischman** [comment]
Do you use Join with Tasker?  It will allow you to send messages between all your devices.

For instance, when my phone enters the "Home Near" region, NodeRed on my computer sends a message to Tasker on my phone with Join.  My message is similar to "ctreischmanlocation=:=home near".  My phone is looking for keyword "ctreischmanlocation" as a location update and "home near" is what it should respond to.  This is similar for everyone in my house with a phone.  I send lots of messages this way.
Once Tasker knows to do something, it performs action "Send Intent" to OwnTracks.  The intent "Action" is "org.owntracks.android.CHANGE_MONITORING" and the "Extra" field will contain "monitoring:2".  "monitoring:2" is the "Move" setting and "monitoring:1" is the "Significant Changes" setting.
When I finally enter the "Home" region, NodeRed sends "ctreischmanlocation=:=home" which sets everything back to normal.

This may have given you more questions than answers.

![Send Intent](https://github.com/hepcat72/node-red-contrib-life360/assets/77683119/2bf76c80-739b-4d0b-b242-03d91d59a303)

By having everyone's location on my phone I was able to make the location clock below, similar to Harry Potter.
![Location Clock](https://github.com/hepcat72/node-red-contrib-life360/assets/77683119/25dc738f-861f-4e87-b361-5abf5bf331e9)

I cannot tell if these attachments are working.

**Comment by ctreischman** [comment]
Wow! They're huge.  Sorry.

**Comment by hepcat72** [comment]
This is great stuff. I'm curious about about reliability. If any of these actions ever fail, what happens?

**Comment by ctreischman** [comment]
Join is great about backlogging the messages in case I don't have phone/internet connection at the time.  Everything seemed reliable except for the Life360, even it was for a while.  To begin with I just left a phone at home with life360 giving notifications and Tasker taking action on the notifications, but the phone batteries would swell an break the phones.  The beauty of the nodered server was that it was always on and didn't have batteries.

If nodered seems to freeze while I'm on vacation, I have power cycled it with a wifi switch.  All my other home switches are zwave controlled by Vera.

Join on the Ubuntu Desktop passes through a Chrome browser extension.  Things get frozen occasionally when there is a chrome update available.  But, I am in no way an IT specialist so there could be a better way of handling that rather than just waiting on it to freeze like me.

**Comment by hepcat72** [comment]
I'm coming from an iOS perspective.  I assume Join is an Android thing?  I'll have to think about a way to automate settings changes... I wonder if the iOS owntracks app has a shortcuts dictionary... Then all I would have to do is figure out how to capture the message from my node red to act on... But still, assuming I can do that, it seems fragile.  Does your phone confirm receipt of the message, and if not confirmed, do you have node-red retry?

BTW, I've never encountered a node-red freeze before. What's going on there? If I ever need to reboot my pi, I issue a `sudo shutdown -r` command. A power cycle is a little risky.

**Comment by hepcat72** [comment]
Yes! OwnTracks has a shortcuts dictionary with a method to "Change Monitoring". Excellent.

**Comment by ctreischman** [comment]
My main freeze, that I think I have solved ended up being with memory or the memory sockets.  Now, Join could stop sending/receiving messages just because of a chrome browser update.  I don't really check on that computer unless I notice a new issue.

I cannot speak to receipts because I am not smart enough.  Any missed messages would seem to be either all or none.

Here is the Join site.  https://joaoapps.com/join/  This guy is super responsive and helpful.  There does appear to be a desktop app for Mac, not sure about phones.  There could be an apple way to pass info between devices after it is sent to a "hub" Mac.

**Comment by hepcat72** [comment]
I think apple has automations pretty locked down in terms of what can be triggered to run without manual confirmation.  I've run into that issue numerous times.  I'll hammer on it when I have time to sit down to it.

**Comment by hepcat72** [comment]
I set out to get OwnTracks working. Found a very easy to follow [tutorial](https://dev.to/webhookrelay/fast--simple-location-tracking-with-node-red-and-owntracks-195o). Now I am receiving location updates on my node-red instance from my iPhone.

Initially, I was reading the OwnTracks booklet, and it is *highly* technical.  Eventually, I decied that reading the booklet was overkill, so I googled and found that tutorial. But one thing I learned while reading the booklet is that OwnTracks on iOS runs differently compared to Android. On Android, it's always running in the background, but on iOS, the process is "killed off". From the booklet:

> OwnTracks on Android runs in the background, all the time, whereas on iOS it is "killed off" by the operating system and woken up every several hundred seconds, in which it gets a teeny tiny time slot to do its thing.

I have not yet set up any functionality for triggering departure/arrival events.  All I've got connected is a debug node that prints my location and it seems to be working.

For anyone else interested in walking through this with me, I took notes.  I skipped the bits about the `worldmap` node.  (I initially tried it, but my Node version is too outdated ATM.)  Here are my notes.  Sorry, I also skipped the webhookrelay setup (because I've already got that going - you can consult the original article above if you need help with that part.)

```
Owntracks notes
Based on: https://dev.to/webhookrelay/fast--simple-location-tracking-with-node-red-and-owntracks-195o

On webhookrelay.com (in browser on computer):
	Clicked Buckets
	Clicked New Empty Bucket
	Entered name "owntracks"
	Clicked Create
	Copied default public endpoint

On iPhone OwnTracks App
	Tap "i" at top left
	Tap "Settings" under "CONNECTION"
	Tap HTTP
	Paste default public endpoint from above in URL field (just above "EXPERT MODE")
	Disable Authentication
	Disable Password
	Enter a UserID (basically, your username - I entered "Rob")
	Enter a DeviceID (basically, the name of your iPhone)
	Back out of Status Info
	Force app to quit (because publish settings wasn't working)
	Restarte app
	Tap "i" at top left
	Tap "Settings" under "CONNECTION"
	Tap "Publish Settings"

On webhookrelay.com:
	Assuming still viewing the "owntracks" bucket page from above
	In the "Requests" area/card, click "REFRESH"
	Confirm that entries have populated with the settings you published from your iPhone
	Click the ellipsis
	Click details
		You should see something like:

{"_type":"dump","configuration":{"username":"user","maxHistory":0,"locked":false,"_type":"configuration","monitoring":-1,"deviceId":"","positions":50,"ranging":false,"cmd":false,"sub":true,"allowRemoteLocation":true,"tid":"","url":"https:\/\/9ur26g3i18vuds1g3pprvt.hooks.webhookrelay.com","ignoreStaleLocations":0,"allowinvalidcerts":false,"usePassword":false,"waypoints":[],"auth":false,"locatorInterval":180,"extendedData":true,"ignoreInaccurateLocations":0,"locatorDisplacement":200,"mode":3,"password":"pass","downgrade":0},"topic":"owntracks\/user\/024EA618-EF19-49D4-85E6-537D7C7550E4\/dump"}

	Click your browser's back button to get back to the bucket page

On iPhone OwnTracks App
	Assuming still viewing the settings page from above
	Enter an arbitrary secret encryption key (i.e. "a password") to encrypt the data that will be sent to webhook relay
	Back out of Status info and go back again to the settings page (not sure if necessary - to "save")
	Tap Publish Settings

On webhookrelay.com:
	Assuming still viewing the "owntracks" bucket page from above
	Click "REFRESH"
	Click the ellipsis on the newest entry (top one)
	Click details
		You should see something like (note the data is now encrypted):

{"_type":"encrypted","data":"xvphPubHelZc39KQ0QdRLXJpAIjJ53pEM55cRf81wv0U\/JQrA9RXndoHbKbyoGZT0F86fTu2kDv+IILDUBxLjWe8RN0YgpGHphbhdth5yHtPJ1kFzqhOaQKvcJJxH5sZG8PwOFSv69JNjtDdGfufIxze4YXIw\/vMX3Bcz9YyWlZMj+Zj1iGcKU7xLEBZfKnSEImqGy5qPlgMKLInJgD2Ozsq52OWe0BliIcjuVQ4aqhC4b0RPMBABCtmUaU7qP6VjIpnYSmMdnVSyCG5i+saMDsiR9iYppSZvSi9NWRPvYUPE7Vn66QXNJ9azfNkceKwnbt0xl5Ix5iHlGNrux6EavRcyPu4O4spON6w8CFP2w0APJgNBnO1N4vmXeRGKY95kgeuf21RF50nel2g1zxhZIePsZyhqlXSq5Uo9Im2ttDORmSjapXI2cLI\/3D7C+wS3qoXqVhkyDMHYEE75TKg6B8SYhu0LXKSp\/7gqMlGaGqXkMjFtwpyJywomASXwRZxS08NTbiMVmCm4EFwsCLd6MsdqEIEjCnjeQ\/5FPXvnqffTpJYdbWot0\/sALjqEeGAE5C4AmHaLhTxj7vpf4+w\/IXO0\/RbgVpnJ3uv2lDuZom3uItP7ZXz+JPKxC5ILVKQBjGyinC0W5ptvs6ZM0gHC7kcfxH39bufrNRj1gzEOtM4tJgnkiu9AiC6IN1MdzmBlX2dG0stA+HRtE9lMtKWqLtfmpj4YZg3zTWCbCwarKPp\/w7K16xIi+36NpgX1NxkUeRgZ07cjiGJn2oaWoJio6tHuHQpOTJFmXBfmCs1UtmC5\/UHQCH8BU7hfSaSRBgku847hYkEYY31EvXxuMk="}

	Click your browser's back button to get back to the bucket page

On your node-red instance:
	I've already set up a webhook relay node, so I'm not going to cover creating a webhookrelay token/secret, but you can go here to create a token: https://my.webhookrelay.com/tokens

	Add the "owntracks" bucket to the bucket list (comma-delimited) in the webhookrelay node's settings
	Click Done
	Add a switch node that takes msg.payload.meta.bucket_name and a has an output for `== "owntracks"`
	Add an node-red-contrib-owntracks node, set its secret to the secret you entered into the OwnTracks iPhone app, and connect it to the switch node's owntracks output
	Connect a debug node to the node-red-contrib-owntracks output
	Click Deploy
	Click the debug tab in the right pane

On iPhone OwnTracks App
	Assuming still viewing the settings page from above
	Tap Publish Settings

On your node-red instance:
	Confirm you get the decrypted owntracks data sent from your phone
```

**Comment by ctreischman** [comment]
I have been using MQTT versus HTTP.  
From the app, it seems like the "locations" may be more reliable than the "transitions".  When I look at the map the blue dot will be exactly on my location, but the device marker could be anywhere along my route from the last couple hours.  And, it doesn't seem to create a leave/enter transition unless the device marker is in the region, so there is much lag even if the app is set to "Move" mode.. 
I don't know how to force a "transition" update.  Maybe my phone is placing the blue dot; I'm not sure.  Utilizing the "locations", I can essentially force the phones to provide an update by changing the mode. 
Maybe the next time my wife leaves I can verify that the "location" shows her correct regions even if the "transition" hasn't occurred yet.  
Another slight complication, as I spoke of earlier about my nested regions, the "location" data lists the regions as an array and requires some additional logic to sort through that instead of the much simpler leaves/enters.
I hope you can understand what I've said.

**Comment by hepcat72** [comment]
Yes, I understand. I've been playing around with it myself, though I'm not as far along as you are. I'm still just doing some basic testing. And yes, I made a similar conclusion as you: the transition events are essentially (or sound like they should be) geofence border crossing events. That's essentially how my Geofency app works and that's why it's unreliable. If that event doesn't happen for whatever reason, it doesn't happen. And I'm not sure that using MQTT would cause that event to buffer and wait for a connection to send because if I understand MQTT correctly (I think I do, but I could be wrong) buffering messages happens at the broker (not on the device). And that transition is attacheded to the location at the time it happened. Does your transition marker appear at or near your geofence border?

BTW, I'm pretty sure your phone is not adding the blue dot. That's owntrack's location data. I think the transition marker is just an event, not a device.

Also, if you create a "card", the yellow and blue map marker will change to whatever image you set. I don't know why they called it a "card".

I also tried changing the "monitoring" setting from significant to move (and every other available value), but in the app, the value never changed from -1, so I don't know if changing the setting is actually working for me. On iOS, changing that setting is controlled via a shortcut action. Another shortcut action I have for OwnTracks is "Send Now", which forces the app to send your location to the server. And I was actually thinking that I may be able to use significant more 100% of the time and simply use a personal automation geofence trigger to force a location update. The only annoying thing is that I found (and reported) a bug. The shortcut actions don't work unless you open the app first. They won't work if the app is in the background.

And after realizing all this and after realizing that my phone dying for the first time ever on Friday, I had an annoying thought... I could use personal automation geofence triggers to roll my own bare bones version of Life360 myself. I already know how to extract my phone's GPS coordinates. I've written a shortcut for that before. The only problem is Apple makes it really laborious to set up a repeating automation, so it's effectively impossible to set up a polling mechanism.  I can use geofence triggers, but using 1 time events isn't reliable.

The other thought I had was I could use plugging and unplugging the charging wire triggers to switch between move and significant modes, but I'm pretty sure I was in significant mode when my phone died last week.

I spent an hour or 2 yesterday trying to mitigate the necessity to open the app to make the automations work by implementing a way to switch back to the previous app automatically after automations finish (I.e. the app has been opened). I actually got that to work yesterday. Thus ended play time, lol.

I just wish Life360 still worked. I'm really ticked at them for shutting the home automation community down. It's so annoying to not have a polished and reliable solution anymore.

**Comment by hepcat72** [comment]
The other thing about MQTT is that I read that (at least for OwnTracks' implementation) it needs a constant active TCP connection to work. That's why on Android, it's constantly running in the background. iOS's operating system doesn't allow all that background activity. Instead, on iOS, the way it was written to work, OwnTracks lets the OS tell it when the phone has moved. OwnTracks can attach an action to that event. I don't know why MQTT is said to benefit from a constant TCP connection, but with the way it works on iOS, it sounds like there's no benefit of using MQTT over HTTP. It doesn't buffer when there's no connection and iOS can't maintain a constant TCP connection anyway.

but,,, I'm no expert, so take that all with a grain of salt. I'm just using what MQTT knowledge I have with what I read from the booklet.

the main benefits of MQTT over HTTP are 1. that the broker buffers messages and delivers them when the recipient has the connectivity to connect to it and 2. it allows you to broadcast messages to multiple recipients. Recipients can subscribe to receive those messages.

And as far as I understand OwnTracks, I'm not sure that there's an effective difference between the 2, other than the code on the server may be a bit bulkier for HTTP because it would have to implement its own buffering to send everyone's location to everyone else, and at least as far as how I intend to use OwnTracks, I don't intend to use it to see where my family members are - just for home automation.

**Comment by ctreischman** [comment]
I was greatly pleased with Life360 while it worked.  I don't usually subscribe to any services, but I did for the reliable home automation geofencing.  I just received a notification yesterday that Life360 would auto-renew on 2/14.  If the authentication cannot be resolved by then, I'm dropping it and dealing with work arounds.

I have been curious about MQTT within the wide world of home automation for some time.  Using it to communicate with OwnTracks pushed me over the edge to figure it out.  It is nice that all household phones can publish to the same topic.  I don't really need to know where everyone is all the time, just whether they are home, near, or far.  Life360 would still work to notify me of where everyone is.

For Alabama, we are essentially snowed in this week so I cannot really test the region updates within the location messages.  It seems weird that the transitions are inconsistent.  My phone knows I'm in the circle.  Why can the app not figure it out?

**Comment by hepcat72** [comment]
Honestly, as far as this node is concerned, the 2FA work around just seems very clunky to me, and it appears that it will be a never ending battle to keep this node working, and it's gotten significantly more difficult to fix it. I may circle back, but I think the path forward for using Life360 with home automations is untenable.  But you're totally right.  It was great while it lasted.  And I still intend to use it for sharing my location with friends and family.  It's WAY more reliable than Apple's `Find My` app.  They got it right.  Unfortunately, they're just "shutting the back door" on us via cloudflare, much the same way the mobile providers did with email-to-text using the same cloudflare service (and that was a "supported" feature! - they just decided to limit use cases to a very pared down scope).

As far as the OwnTracks transition events, I don't think they are useful, honestly.  Every location event has an array of waypoints that you're in.  I think that the way to handle leave/arrive events is the same way `node-red-contrib-life360` did.  Just keep track of each user's current "place"(/waypoint), and when that changes, trigger an arrival or departure event.

I just looked at my OwnTracks logfile about 15 minutes ago for the first time after lunch.  I'd taken 2 trips since I started logging events a couple days ago.  I then realized that the json doesn't have a time, so I just added a timestamp to the log entries to give me a better sense.

**Comment by StoricU** [comment]
Is there a way around the 2FA crap?
Like creating a new account or something?

After testing OwnTracks on android for a while now, I have concluded that it is not satisfactory enough.
The battery drain is just to big. :/
Life360 was just working so good, nothing will ever be able to match 😂 🙈

**Comment by hepcat72** [comment]
I have heard of a couple workarounds. Yes, you may be able to create a new account and not add a phone number (though I wouldn't put it past them to start requiring a phone number).

The other option is to add a 2FA interface in the node (request a code be texted to you and then submit it in the node after you get it via text). I haven't implemented that (yet) mainly because it stops working when you open Life360 on your phone and no one has figured out a good solution for that IMO. Though some people have added a *second* account using a google voice number for that purpose - to avoid what people think is intended to be a "number of devices limitation" requirement - though it's not understood well because others have noted that their iPad and phone both work with Life360 *with* a phone number - so I think it's something other than a number of devices limitation.

What I've implemented (on the `auth2024` branch described above *should* work if you don't have a phone number in your account, from what I understand, but I haven't tested it. If you test it by adding an account without a phone number, let me know what happens.

**Comment by hepcat72** [comment]
Another option (actually, I'm not sure it's available on android) is an app called Geofency. I added webhook calls to it for one of my critical automations. It's based on boundary crossing events, which is not reliable, and it's just tracking me currently, but it also stores data that you can synch with iCloud. I turned on my synch the other day so I could look and see if I can use it for better location tracking. I haven't dug through it yet because it was synching years of data after I turned it on and I gave up waiting.

**Comment by hepcat72** [comment]
I sat down to see if I could create a Life360 account without a phone number and it appears to be impossible.  They require a phone number, so it looks like that option was a wild goose chase. I looked in another HS issue about their Life360 integration failing and I noted they added a `wontfix` label to it last week. I think automations triggered by Life360 is dead in the water, though I didn't check to see what HA officially says about it. I was just reading through 1 issue.

I did see an Android user complain about OwnTracks, saying it randomly stops sending updates every few days. I also reached out to the Geofency author to see what thoughts he had about continuous tracking data.  I'm going to continue playing with that and OwnTracks.

**Comment by hepcat72** [comment]
Annndddd.... I just peeked at my owntracks log.  I'd added a timestamp on the 16th and have taken multiple trips since then. Not a single new entry in the log. :(

**Comment by hepcat72** [comment]
Oh, never mind. Somehow, it switched owntracks to "Quiet" mode. I must've done that by acident.

**Comment by hepcat72** [comment]
I know that many of you are on Android, but (and please keep reading, I'll cover Android) I think it's worth noting that I just learned that iOS 17 newly allows location-based automations to run without forcing you to manually tap a notification to allow it to run.  I could for instance, roll my own simple webhook via webhook relay and track presence (of my wife and I) manually, albeit with the same fragile 1-time call for arrival/departure events.  However, I could also use the OwnTracks "Send Location" action.  I know that's a big-ask for larger families, but you can at least share shortcuts that perform the actions.  Each member however would have to set up the trigger.

I imagine there must also be mechanisms to both trigger Android-built-in automations based on location and force OwnTracks to send your location to your server.  OwnTracks has a "manual" mode and I expect that's why - so you only track location using built-in functionality, thus no battery drain.

You could set up automations for other locations, but IMO, it's too complicated of a synch issue to put in that effort, so I would probably set up "waypoints"(/"places") on the server and only use the sending phone's geo-coordinates.  If you combine that with triggers like "starting"/"ending" a drive, you could at least be able to set up some laggy automations for places other than home.

Personally, I was thinking of creating automations to toggle OwnTracks' "move" mode when a charger is connected/disconnected.  But I'll have to wait until my iPhone finishes its update to see if what I expect the new automation functionality is, is actually what it is.

**Comment by ctreischman** [comment]
Android has the Tasker app for these automations.  So far I have OwnTracks placed in "move" while charging as you stated.  But also, when my phone connects or disconnects from an auto's bluetooth, I put it into "move" for just 5 seconds.  Not sure about IOS, but you can also trigger on wifi connected/disconnected or wifi near for 5 seconds.  The SSIDs of your offices, school, church, or doctors can be found to signal the switch.  I think all AMC theaters have the same SSID.  That is how I silence my phone automatically at each one.  

Home is the most important to be reliable and timely.  I cant really wait for my wifi to connect or my car to turn off to open the garage door.  I guess I could figure out a neighbor's SSID to open my door or set to "move" when passing their house.

**Comment by hepcat72** [comment]
I've been using OwnTracks in significant mode for a few days now and I am not liking the the battery drain at all. The more I think about it, the more I'm liking the idea of simply sending my (and my wife's) geo-coordinates in a handful of automations (arrive/leave home [maybe also work], connect to wifi, complete a drive, etc). I suspect that there's probably a node to compute distance between a 2 sets of geo-coordinates and if there's not, the math is pretty straightforward.  I can just obtain coordinates for the places where I want automations to trigger and keep them in a list.  March through the list and output the place or places where a person is.  If it changed from their previous place, trigger an event. I'm actually kind of geeking out about implementing it.

**Comment by hepcat72** [comment]
Don't know if it's the best, but this node will calculate the distance between 2 points: https://flows.nodered.org/node/node-red-contrib-rollun-geolib

**Comment by ctreischman** [comment]
Even in significant changes or move, it isn't working right. My wife and I are both at work and it still indicates that we are home. In OwnTracks I clicked exit app and then my location updated. Wife's is still the incorrect. So for some reason it is getting disconnected. 
Once I have some time at home, I'll also try the library you have suggested.

**Comment by hepcat72** [comment]
I haven't tried it myself yet, but I did install and look at it. It's only 1 node that outputs the distance between 2 pairs of GPS coordinates (longitude and latitude).  The interface is a bit awkward.  And there's not an option for feet or meters.  I'll probably just look at the code and create a function node to do it.

**Comment by ctreischman** [comment]
You should check out node-red-node-geofence.  It is doing well with comparing a lat/lon to a geofence.  It gives a true/false for inside/outside.  You just need to come up with your own way for your phone to send lat/lon to node red.  On android, Tasker is pulling my location and sending it to NoneRed with Join.  I am just having to come up with different triggers to manually send my location.  OwnTracks is not seeming reliable.  It seems to randomly get disconnected and doesn't reconnect without exiting the app.

**Comment by StoricU** [comment]
The problem for me is not the trigger part. It is the part about getting the location reported without draining the battery.
I'm really grumpy that Life360 changed their policy, since they did it really well.. 😏

**Comment by hepcat72** [comment]
OK.  I was playing around this morning.  I'm still in the very beginning stages of rolling my own geo-fence automations, using Life360 as a model.  I am going to be creating iOS shortcuts to create a Person, Place, and Group(/"Circle"); and one to send location.  I'll be storing this Data in Node-RED global variables, but one thing I realized I would like to have, is the ability to retrieve the data (person, group, place) on my phone so that I would at least be able to debug.  And I want to be able to explicitly add a person or place to a group, so I need to at least be able to retrieve the groups.  (I'll only have 1 group, e.g. "home automations", and one place, "home", but I want to set things up so it can grow - and I like how Life360 handled that.)

So I set about being able to request data from my pi via webhook relay and I was able to get that configured by following [this tutorial](https://webhookrelay.com/blog/2020/03/02/responding-to-api-calls-on-nodered/).

And I created a test webhook call to my webhook relay endpoint/bucket I created for dynamic responses:

<img width="1176" alt="Node-RED raspberrypi local - Screen Shot 2024-01-27 at 12 23 PM" src="https://github.com/hepcat72/node-red-contrib-life360/assets/2300532/b5483898-96f0-418e-8706-7d2afd768c24">

I didn't use function nodes like in the tutorial and instead of using the complex weather example, I just set a static template test message that simply says "This is my custom response.".

Note, I have a separate flow for incoming webhook relay messages:

<img width="346" alt="whrmessages" src="https://github.com/hepcat72/node-red-contrib-life360/assets/2300532/56ea80f5-e89d-4fa5-8a7a-0c23518efd26">

The two flows link to one another.

I'm also working on a shortcut to encrypt my location data, because when my phone sends my location, it's clear text, and that's a security issue.
